### PR TITLE
Maas 4215 merge worker with alerter

### PR DESCRIPTION
**Pull request by tetianakh**
Merged alamo alerter into alamo worker.

Modified alamo alerter to use asyncio to connect to cassandra db and to execute http requests.

### CODE REVIEW

File: alamo_worker/manage.py [hunk 1]
```diff
@@ -64,14 +73,21 @@ def close(self):
...
+    async def send(self, result):
         """Process check with received result to alerter."""
 
         logger.debug("Enqueuing result: %s", result)
         self._send_result_to_statsd(result)
-        self.queue.enqueue('alamo_alerter.job.analyze',
-                           args=(result,),
-                           result_ttl=settings.WORKER_RESULT_TTL)
+        try:
+            await self._alerter.execute(result)
```
Comment (deejay1) [inline alamo_worker/manage.py]: won't this fail horribly if `self._alerter = None` ?

File: requirements.txt [hunk 2]
```diff
@@ -1,13 +1,22 @@
...
-pyzmq==16.0.3
-redis==2.10.6
-rq==0.8.2
-statsd
-pytz>=2017.3
+aiocassandra==2.0.1
 aiohttp==2.3.2
+alamo-common>=0.4.5
+cached-property==1.3.0
+cassandra-driver==3.11.0
```
Comment (deejay1) [inline requirements.txt]: aiocassandra 2.0.1 requires cassandra-driver==3.13.0

File: requirements.txt [hunk 3]
```diff
@@ -1,13 +1,22 @@
...
-stevedore==1.27.1
-alamo-common>=0.4.5
+python-dateutil==2.6.1
+pytz>=2017.3
+pyzmq==16.0.3
+raven==6.3.0
+redis==2.10.6
+requests==2.18.4
+retry==0.9.2
+rq==0.8.2
```
Comment (deejay1) [inline requirements.txt]: is this still needed ?

File: requirements.txt [hunk 4]
```diff
@@ -1,13 +1,22 @@
...
 cchardet==2.1.1
+pbr
 pydruid==0.3.1
 python-consul==0.7.2
-stevedore==1.27.1
-alamo-common>=0.4.5
+python-dateutil==2.6.1
+pytz>=2017.3
+pyzmq==16.0.3
+raven==6.3.0
```
Comment (deejay1) [inline requirements.txt]: can we bump this req when we're at it ?

File: alamo_worker/http_client.py [hunk 5]
```diff
@@ -39,9 +39,9 @@ def __init__(self, timeout):
         with Timeout(self._timeout):
             async with ClientSession(connector=connector) as session:
                 async with session.request(method, **call_kwargs) as response:
+                    await response.read()
```
[the remark this answers is not in the document]
Comment (tetianakh) [reply alamo_worker/http_client.py]: Someone may catch the exception and read the content.

**Comment by ch40s3k** [comment]
PEP-8 + merge conflicts

File: alamo_worker/alerter/alerter.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+class AlamoAlerter(object):
+    driver = None
+
+    def __init__(self, driver, notification, statsd):
+        self.driver = driver
+        self.notification = notification
+        self.statsd = statsd
+        self._preprocessor = AlertPreprocessor()
+
+    @stats.timer(hostname=True)
```
Comment (pnuckowski) [inline alamo_worker/alerter/alerter.py]: for `async` method maybe we should use `aiostats` ?

File: alamo_worker/alerter/job_configuration.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            '.'
+        )
+        try:
+            module = importlib.import_module(mod_path)
+
+            module = getattr(module, cls_name, None)
+            assert module, (
+                'Could not import `{}` from `{}`.'.format(cls_name, mod_path)
+            )
+            return module()
```
Comment (pnuckowski) [inline alamo_worker/alerter/job_configuration.py]: `return` ?

File: alamo_worker/alerter/job_configuration.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+from alamo_worker.conf import settings
+
+logger = logging.getLogger(__name__)
+
+
+class JobConfiguration(object):
+    """Simple singleton class to read alerter config file only once.
+    """
+    __instance = None
+
```
Comment (pnuckowski) [inline alamo_worker/alerter/job_configuration.py]:
```python    
    __configured = False
    def __new__(cls, *args, **kwargs):
        if not hasattr(cls, '_inst'):
            cls._inst = super().__new__(cls, *args)
        return cls._inst
```
more readable

File: alamo_worker/alerter/alerter.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+logger = logging.getLogger(__name__)
+
+
+class AlamoAlerter(object):
+    driver = None
+
+    def __init__(self, driver, notification, statsd):
+        self.driver = driver
+        self.notification = notification
+        self.statsd = statsd
```
Comment (pnuckowski) [inline alamo_worker/alerter/alerter.py]: `statsd` is not used

File: alamo_worker/alerter/job_configuration.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+from statsd import StatsClient
+
+from alamo_worker.alerter.backend import AsyncCassandraDriver
+from alamo_worker.conf import settings
+
+logger = logging.getLogger(__name__)
+
+
+class JobConfiguration(object):
```
Comment (pnuckowski) [inline alamo_worker/alerter/job_configuration.py]: I think this class should be dropped and all it's content moved to `AlamoAlerter`.

Alerter rely on driver, notification and preprocessor, so it's natural that all those properties are initialized there.

File: alamo_worker/manage.py [hunk 11]
```diff
@@ -26,26 +27,36 @@ class WorkerManagement(object):
     soft_exit = False
     open_monitor_socket = False
     # scheduler connections
-    scheduler_pool = None
+    __scheduler_pool = None
+    __queue_size_metric = 'queue_size'
```
Comment (pnuckowski) [inline alamo_worker/manage.py]: Maybe it should include worker FQDN?
`"{}.queue_size".format(settings.WORKER_FQDN)`

Comment (deejay1) [reply alamo_worker/manage.py]: with `.` replaced by `-`

File: alamo_worker/alerter/tests/test_alerter.py [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        self.trigger = self.check.triggers[0]
+        self.notification = Mock()
+        self.alerter = AlamoAlerter(
+            mock,
+            self.notification,
+        )
+
+    def test_retrieve_cached_result_by_rebuilding_cache(self):
+        self.alerter.driver.get_result = AsyncMock(return_value=[])
+        result = run_async(self.alerter.get_cached_result(
```
[the remark this answers is not in the document]
Comment (pnuckowski) [reply alamo_worker/alerter/tests/test_alerter.py]: Maybe in next iteration someone will do it :)

**Pull request merged**